### ISSUE TITLE
Switch to ES6 Promise

### ISSUE THREAD

**Issue by matthewma7** [girder/girder#2068]
@brianhelba has [suggested](https://github.com/girder/girder/issues/2030) using jQuery Deferred(Promise) instead of ES6 standard Promise. However, as a heavy user of standard Promise in the past few years, I am a strong advocate for standard Promise. 
Brian and I tried to convert some of my code that uses standard Promise to use jQuery promise last week, aiming to see if such conversion would be smooth, but we found some drawbacks of jQuery Deferred. Brian thinks I should create a post to promote standard Promise.

Here are my reasons to use standard ES6 Promise start now and transit existing usage of jQuery Deferred(Promise) to standard ES6 promise with girder 3.

- Standard promise is more popular and widely accepted than jQuery deferred ever was and will be. 
  -  Node.js application uses the standard Promise, but jQuery generally is not used in Node.js
  -  Third party libraries or frameworks are developed around standard Promise for interoperability (such as Bluebirdjs)
  -  Native APIs (such as [MediaDevices.getUserMedia()](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia))
- Promise is not only the standard but a language feature with the introduce of async and await ([demo](https://jsfiddle.net/jfmorio/hxz00roq/))
- jQuery Deferred and Promise concepts are confusing, for people who know neither, standard Promise is easier to learn
- When is awkward to use (a lot usage of ... spread)
- jQuery when is return data structure is "**broken**". Unlike Promise.all(), when() returns different structure for zero, one or more arguments.
- jQuery Deferred(Promise) is not interoperable with standard Promise chain and await keyword because it returns more than one result sometimes ($.ajax.then, $.when) and all arguments except the first will be stripped out.
- [Example for the above three points](https://jsfiddle.net/jfmorio/LLaqf7sa/2/)
- ES6 promise makes girder look better. Nowadays "jQuery" makes people raise an eyebrow, but there won't be any question if you tell people you use ES6 promise.

All the above reasons make usages of jQuery deferred, promise technical debts. The only two drawbacks I can think of if we start using standard Promise are 
- ~~some mixed use of jQuery and standard Promise before moving to standard Promise~~
- A polyfill is needed for IE 11
- breaking change with girder 3

**Comment by cjh1** [comment]
:+1: For sticking with the standard/language feature.

**Comment by sebastienbarre** [comment]
👍 Same here. Stick with jQuery might bite back...

**Comment by zachmullen** [comment]
Let's attempt to quantify the cost of this effort, and its benefits.

**Comment by matthewma7** [comment]
If we say starting using standard promise now, the cost is a polyfill, existing code and plugins won't break because of that, the benefit is productivity with standard Promise and debt free, future proof Promises. After all, there are reasons to switch from jQuery Deferred to standard Promise, but there won't be for switching Promise back to jQuery.

If we say changing the existing usage of jQuery deferred to standard Promise, I took a look at current Girder core repository, there are 140 uses done(), 100 for .fail(), and 16 for always(), only 7 places uses $.when(). Almost all of them are chaining of a restRequest(). So, the overall of usage of jQuery deferred is not logically complex. One of the reasons behind the situation is, as suggested by @brianhelba, and I totally agree, is that we are not using Promises as much as we should be. 
The usage in Minerva is a similar picture there.
I guess the usages in other plugin are similar. Aside from other indirect costs, the actually work to change existing code is actually quite straightforward. 

On the benefit side, as suggested by Brian and as I am doing, more Promise and complex Promise usage should be encouraged in favor of other async mechanisms, which would be much harder to change than simple .done(). (Brain and I spent over an hour trying to change the promises in item preview PR  to jQuery promises and still failed to do so due to the $.when drawbacks). The early we start using standard Promise, the less debt there will be, the fewer hours will be spent on $.when quirks. Since almost all existing usage of Promise is for restRequest(), I think we could even say we are only at the beginning phase of **real** Promise use. Starting using standard Promise makes more sense because of this.

Overall, I think the cost is fixed and the change is straightforward at this moment to switch existing usage Promise to standard. The benefit is long term and will benefit all future uses of Promise within girder or overall web development.

**Comment by brianhelba** [comment]
I'm not yet ready to stake out my position on the technical merits of each Promise library (ES6 vs jQuery3).

However, consistency within Girder is critical, both across the API surface and within Girder's code conventions. If we move forward with this change, it should entail using ES6 to instantiate all Promises and to wrap all `restRequest` returns in ES6 Promises. Unlike the change to use arrow (`=>`) functions from `_.bind`, this is not a change we can make piecemeal: Promises are the mechanism that Girder functions use to communicate with each other, so this communication mechanism must be uniform and predictable (for both public and private methods). Even within a single function, we must be consistent, especially given that Promises are a programming mechanism that some developers are less familiar with.

So, let's be sure to account for the fact that this change would have to occur as part of a new major version release (i.e. Girder 3.0) and plan accordingly.

**Comment by matthewma7** [comment]
I would say that what's more inconsistent and unpredictable is the number of argument in jQuery then() and the behavior of $.when() :sweat_smile:

**Comment by brianhelba** [comment]
@matthewma7 To help with your difficulties in using `$.when` in a nested way, note that [in your example](https://jsfiddle.net/jfmorio/LLaqf7sa/1/), you can change line 22 from:
```
return $.when(...$promises);
```
to
```
return $.when(...$promises).then((...results) => results);
```
to get the exact same result as `Promise.all`.

Accordingly, as brought up by @jbeezley , if the structure of `$.when` is causing you undue development pain, you could easily create a helper:
```
function _whenAll (promiseArray) {
    return $.when(...promiseArray).then((...results) => results);
}
```
and use it wherever you need the exact behavior of `Promise.all`. I would probably keep it private (prefixed by underscore), to keep us from having to deprecate / support it for too long into the future.

**Comment by matthewma7** [comment]
@brianhelba, I wasn't saying there can't be workarounds, but workarounds oppose the point you made about [Use promises more consistently and thoroughly](https://github.com/girder/girder/issues/2030). You are promoting the use of $.when in the post. 
I believe it will be harder to be consistently and thoroughly with a patched version that no one knows.

**Comment by jeffbaumes** [comment]
The proposal that I'll put on the table, since after discussion it appears more clear that we cannot move to ES6 Promise until Girder 3:

* Use jQuery promises (including `$.when` and/or a `_whenAll` utility described by @brianhelba, even though this is a non-elegant solution) on branches to be merged pre-Girder 3.
* Make an issue tagged for Girder 3, in which all code will be moved to ES6 Promise in one fell swoop.

Everyone likely realizes that starting to use ES6 Promise now would avoid more effort converting later, but backwards compatibility and consistency until Girder 3 requires a route like this.

**Comment by matthewma7** [comment]
@jeffbaumes, thanks for the input. Moving to standard Promise eventually makes sense. 

@all, I would like to promote the utility function to a real utility function in girder because there are and will be multiple places needs it. And it potentially would help switching to ES6 promise in one go.

**Comment by brianhelba** [comment]
@matthewma7 While I still don't oppose a migration (I'll remain undecided for now), I do want to comment:

---

I generally am of the opinion that excessive use or reliance on `$.each`, `Promise.all`, etc. on the client side is a code smell (server-side is a different story), so I'm suspicious of too many uses of this `_whenAll` utility.

If a single operation takes many API requests to complete, it should pipeline the processing and incremental rendering of data as it's received. This has the benefit of increasing performance (as you spend your I/O waiting time doing useful processing, as opposed to deferring all of the processing until the very end). It also should make your pages more responsive, giving users something to view and interact with as early as possible, then filling in additional details or content as it arrives (users perceive time spent waiting vs. time spent reading something very differently).

Furthermore, your main difficulty with `$.each` was coming from a place where you were using it in a nested loop. If you've got `O(n*m)` asynchronous requests, *all of which* need to complete before you do some final steps, I would be concerned about the scalability of the code architecture.

In the case that you truly *need* multiple requests to get any sensible results at all, adding a new API endpoint should probably be considered.

Of course, like all "code smell" issues, there are legitimate exceptions (e.g. displaying the message "done loading" or something), so maybe I'm totally wrong. :fearful: 

---

What you're doing in [your test branch](https://github.com/girder/girder/tree/use_standard_promise) is not going to be adequate. This code:
```
Promise.prototype.done = Promise.prototype.then;
Promise.prototype.fail = Promise.prototype.catch
Promise.prototype.always = function (func) {
    return this.then(func, func);
}
```
treats these function as aliases, which they are definitely not. `.done`, `.fail`, and `.always` return the original promise, whereas `.then` returns a new Promise. Of course, this could be fixed by a slightly longer patched implementation.

More importantly, even in jQuery3, `.done`, `.fail`, and `.always` execute synchronously at the time of attachment if the Promise is already resolved. This generally doesn't matter in production, but can have significant implications in a test environment, as evidenced by the multiple failing tests that had to be fixed to get #2063 consistently passing. Of course, in this context, failing tests are "good", as they reveal improperly written tests, but it is still more work to refactor them.

The best strategy here may be to skip over ES6 Promises and use [Bluebird.js](http://bluebirdjs.com/docs/getting-started.html), as it uses same Promise instantiation syntax as ES6, while exposing `.tap` and `.tapCatch` methods that more directly map to the functionality of `.done` `.fail`, and thus may make refactoring easier (as otherwise, all of our `.then` methods are expected to explicitly return something).

**Comment by jeffbaumes** [comment]
@girder/developers here's the Promise discussion. Looks like the proposals are to move to ES6 Promise or Bluebird.js in Girder 3. I'm not seeing any argument for keeping the status quo, though any discussion is welcome.

**Comment by matthewma7** [comment]
UPDATE: I added a new point and demo for the benefits of ES6 promise.

**Comment by mgrauer** [comment]
Mooted by dint of Backbone client jettison.
